belt-dwp/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+//!
+//! This crate has an optional `alloc` feature which can be disabled in e.g.
+//! microcontroller environments that don't have a heap.
+//!
+//! The [`AeadInPlace::encrypt_in_place`] and [`AeadInPlace::decrypt_in_place`]
+//! methods accept any type that impls the [`aead::Buffer`] trait which
+//! contains the plaintext for encryption or ciphertext for decryption.
+//!
+//! Note that if you enable the `heapless` feature of this crate,
+//! you will receive an impl of [`aead::Buffer`] for `heapless::Vec`
+//! (re-exported from the [`aead`] crate as [`aead::heapless::Vec`]),
+//! which can then be passed as the `buffer` parameter to the in-place encrypt
+//! and decrypt methods:
+//!
+#![cfg_attr(all(feature = "os_rng", feature = "heapless"), doc = "```")]
+#![cfg_attr(not(all(feature = "os_rng", feature = "heapless")), doc = "```ignore")]
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! use belt_dwp::{
+//!     aead::{AeadInPlace, AeadInPlaceDetached, KeyInit, heapless::Vec},
+//!     BeltDwp, Nonce
+//! };
+//!
+//! let key = BeltDwp::generate_key().unwrap();
+//! let cipher = BeltDwp::new(&key);
+//! let nonce = Nonce::try_from(&b"unique nonce1234"[..]).unwrap(); // 128-bits; unique per message
+//!
+//! let mut buffer: Vec<u8, 128> = Vec::new(); // Note: buffer needs 16-bytes overhead for auth tag
+//! buffer.extend_from_slice(b"plaintext message");
+//!
+//! // Encrypt `buffer` in-place, replacing the plaintext contents with ciphertext
+//! cipher.encrypt_in_place(&nonce, b"", &mut buffer)?;
+//!
+//! // `buffer` now contains the message ciphertext
+//! assert_ne!(&buffer, b"plaintext message");
+//!
+//! // Decrypt `buffer` in-place, replacing its ciphertext context with the original plaintext
+//! cipher.decrypt_in_place(&nonce, b"", &mut buffer)?;
+//! assert_eq!(&buffer, b"plaintext message");
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! Similarly, enabling the `arrayvec` feature of this crate will provide an impl of
+//! [`aead::Buffer`] for `arrayvec::ArrayVec` (re-exported from the [`aead`] crate as
+//! [`aead::arrayvec::ArrayVec`]).
+
+use aead::consts::{U8, U16, U32};
+pub use aead::{self, AeadCore, AeadInPlace, Error, Key, KeyInit, KeySizeUser};
+use aead::{AeadInPlaceDetached, PostfixTagged};
+use belt_block::cipher::{Block, BlockCipherEncrypt, KeyIvInit, StreamCipher};
+use belt_block::{BeltBlock, belt_block_raw};
+use belt_ctr::BeltCtr;
+use universal_hash::UniversalHash;
+
+use crate::{
+    ghash::GHash,
+    utils::{from_u32, to_u32},
+};
+
+/// Nonce type for [`BeltDwp`]
+pub type Nonce = aead::Nonce<BeltDwp>;
+
+/// Tag type for [`BeltDwp`]
+pub type Tag = aead::Tag<BeltDwp>;
+
+mod gf;
+mod ghash;
+mod utils;
+
+/// T from the STB 34.101.31-2020
+const T: u128 = 0xE45D4A588E006D363BF5080AC8BA94B1;
+
+/// Belt-DWP authenticated encryption with associated data (AEAD) cipher, defined in
+/// STB 34.101.31-2020
+pub struct BeltDwp {
+    key: Key<BeltBlock>,
```

Comment:
It's better to keep initialized `BeltBlock` in `BeltDwp` and pre-computed `r` and `s` constants instead of the raw key. Note that you should be able to create and use an instance of `BeltCtr<&BeltBlock>`.

Reply:
I didn't really understand how to do this, because s and r are created at the stage with the transfer of Nonce. That's why I did it this way.

Reply:
Ah, true. Then just use initialized `BeltBlock` instead of the raw key.

Reply:
You should use `BeltCtrCore::inner_iv_init(&self.cipher, &nonce)`.

Reply:
Looks like for that, I need to copy BeltBlock on each operation. Is it fine?

Reply:
No, you can create a `BeltCtr` instance using reference to the block cipher (note the `&self.cipher` part). In other words, you will use `BeltCtr<&BeltBlock>`, not `BeltCtr<BeltBlock>`.

Reply:
Nvm. Just removed Cipher. Don’t thinks it is needed.